### ISSUE TITLE
wranglerjs is too rigid in what version of webpack it uses to build projects

### ISSUE THREAD

**Issue by ashleymichal** [cloudflare/wrangler#1558]
from @1000hz : the parser webpack 4 uses (acorn) doesn't support optional chaining / nullish coalescing operators. wranglerjs uses its own version of webpack (4.x). TypeScript target `ES2020`/`ESNext` doesn't transpile these operators
webpack 5 (which has been in beta forever) uses a newer version of acorn with support for these operators
but there's no way to override what version of webpack wrangler will use

**Comment by 1000hz** [comment]
I think the long-term solution to this problem is to have wrangler defer building to project scripts, i.e.
if there's a package.json file
```
npm run wrangler:build
npm run build
```
or if there's a cargo.toml file
```
cargo run wrangler:build
cargo build
```
and only fallback to using a bundled version of webpack if none of these exist.

**Comment by caass** [comment]
Supporting arbitrary scripts seems like a really good idea actually, and would allow people to choose their own JS bundling tools (e.g. rollup, etc). If nobody else is working on this i'm going to poke around and see if it's something that's within the scope of my rust knowledge to work on. No promises!!

**Comment by Gowee** [comment]
It turns out that wrangler installation comes with `wrangler-js` which has `webpack 4.33.0` built-in. 

I can build my project successfully with `webpack 5.x`. But `wrangler dev` or `wrangler build` invokes its internal old version regardless. Is there any way to let `wrangler` use my manually built output?

**Comment by xtuc** [comment]
Once https://github.com/cloudflare/wrangler/pull/1677 is merged, we plan to support any version of webpack

**Comment by xortive** [comment]
closing since this is addressed by #1677, which will be going out as part of an RC in the near future
